keras/engine/topology.py

```diff
@@ -1690,6 +1690,14 @@ def build_map_of_graph(tensor, finished_nodes, nodes_in_progress,
             # If the depth is not set, the node has no outbound nodes (depth 0).
             depth = nodes_depths.setdefault(node, 0)
 
+            # Update the depth of the corresponding layer
+            previous_depth = layers_depths.get(node.outbound_layer, 0)
+            # If we've seen this layer before at a higher depth, we should use that depth instead
+            # of the node depth.  This is necessary for shared layers that have inputs at different
+            # depth levels in the graph.
+            depth = max(depth, previous_depth)
```

Comment:
Hmm, I might need to also update `node_depths[node]` here...  I didn't test the node_depth calculation.